azkaban-common/src/main/java/azkaban/scheduler/MissedSchedulesManager.java

```diff
@@ -0,0 +1,199 @@
+      }
+      return true;
+    } catch (RejectedExecutionException e) {
+      LOG.error("Failed to add more missed schedules tasks to the thread pool", e);
+      return false;
+    }
+  }
+
+  public void start() {
+    // TODO: add metrics to monitor active threads count, total number of missed schedule tasks, etc.
+    // Create the thread pool
+    this.missedSchedulesExecutor = this.missedSchedulesManagerEnabled
+        ? Executors.newFixedThreadPool(threadPoolSize, new DaemonThreadFactory("azk-missed-schedules-task-pool"))
+        : null;
+    if (this.missedSchedulesManagerEnabled) {
+      LOG.info("Missed Schedule Manager is ready to take tasks.");
+    } else {
+      LOG.info("Missed Schedule Manager is disabled.");
+    }
+  }
+
+  public boolean stop() throws InterruptedException {
+    if (this.missedSchedulesExecutor != null) {
+      this.missedSchedulesExecutor.shutdown();
+      if (!this.missedSchedulesExecutor.awaitTermination(30, TimeUnit.SECONDS)) {
+        this.missedSchedulesExecutor.shutdownNow();
+      }
+    }
+    return true;
+  }
+
+  /**
+   * A MissedSchedule Task is a unit that store the critical information to execute a missedSchedule operation.
+   * A Task is capable to
+   * 1. notice user
+   * 2. execute back execution if enabled feature flag (TODO)
+   * when a missed schedule happens.
+   * */
+  public static class MissedSchedulesOperationTask implements Runnable {
+    private static final Logger LOG = LoggerFactory.getLogger(MissedSchedulesOperationTask.class);
+    private final Emailer emailer;
+    // emailRecipients to notify users the schedules missed
+    private final List<String> emailRecipients;
+    // the email body
+    private final String emailMessage;
+    // back execute action
+    // if user disabled the config, the action could be null.
+    private ExecuteFlowAction executeFlowAction;
+
+    public MissedSchedulesOperationTask(List<Long> missedScheduleTimes,
+        Emailer emailer,
+        List<String> emailRecipients,
+        boolean backExecutionEnabled,
+        ExecuteFlowAction executeFlowAction) {
+      this.emailer = emailer;
+      this.emailRecipients = emailRecipients;
+      this.executeFlowAction = backExecutionEnabled ? executeFlowAction : null;
+      final String missScheduleTimestampToDate = formatTimestamps(missedScheduleTimes);
+      MessageFormat messageFormat = new MessageFormat("This is autogenerated email to notify to flow owners that"
+          + " the flow {0} in project {1} has scheduled the executions on {2} but failed to trigger on time. "
+          + (backExecutionEnabled ? "Back execution will start soon as you enabled the config." : ""));
+
+      this.emailMessage = messageFormat.format(
+          new Object[]{executeFlowAction.getFlowName(), executeFlowAction.getProjectName(), missScheduleTimestampToDate});
+    }
+
+    private String formatTimestamps(List<Long> missScheduleTimestamps) {
```

Comment:
Can we use a utility method in TimeUtils class for the formatting?

Reply:
Sorry, I should have described more. I meant checking TimeUtils methods like `formatDateTimeZone` or `formatDateTime` to see if we could use those instead of writing the formatting logic here. Didn't mean to create a utility method for the strings work.

Reply:
sure I changed to leverage formatDateTime.